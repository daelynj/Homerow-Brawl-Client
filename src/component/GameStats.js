import React from "react";

function GameStats(props) {
  function getLettersTyped() {
    return props.words.join("").length;
  }

  function calculateAccuracy() {
    var letters = getLettersTyped();
    let accuracy = (
      ((letters - props.incorrectLetters) / letters) *
      100
    ).toFixed(1);

    return accuracy > 0 ? accuracy : 0;
  }

  //https://www.speedtypingonline.com/typing-equations
  function calculateWPM() {
    let allTypedEntries = getLettersTyped();
    let averageWordLength = 5;
    let timePlayed = props.finishTime / 60;
    let grossWPM = allTypedEntries / averageWordLength / timePlayed;

    return Math.round(grossWPM);
  }

  return (
    <div>
      <div>{"words typed: " + props.words.length}</div>
      <div>{"WPM: " + calculateWPM()}</div>
      <div>{"time: " + props.finishTime + " seconds"}</div>
      <div>{"mistakes: " + props.incorrectLetters}</div>
      <div>{"accuracy: " + calculateAccuracy() + "%"}</div>
    </div>
  );
}

export default GameStats;
