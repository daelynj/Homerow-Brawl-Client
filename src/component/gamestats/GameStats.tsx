import * as React from "react";

interface Props {
  words: String[];
  incorrectLetters: number;
  finishTime: number;
}

export const GameStats = (props: Props) => {
  const getLettersTyped = () => {
    return props.words.join("").length;
  };

  const calculateAccuracy = () => {
    let letters: number = getLettersTyped();
    let accuracy: any = (
      ((letters - props.incorrectLetters) / letters) *
      100
    ).toFixed(1);

    return accuracy > 0 ? accuracy : 0;
  };

  //https://www.speedtypingonline.com/typing-equations
  const calculateWPM = () => {
    let allTypedEntries: number = getLettersTyped();
    let averageWordLength: number = 5;
    let timePlayed: number = props.finishTime / 60;
    let grossWPM: number = allTypedEntries / averageWordLength / timePlayed;

    return Math.round(grossWPM);
  };

  return (
    <div>
      <div>{"words typed: " + props.words.length}</div>
      <div>{"WPM: " + calculateWPM()}</div>
      <div>{"time: " + props.finishTime + " seconds"}</div>
      <div>{"mistakes: " + props.incorrectLetters}</div>
      <div>{"accuracy: " + calculateAccuracy() + "%"}</div>
    </div>
  );
};
