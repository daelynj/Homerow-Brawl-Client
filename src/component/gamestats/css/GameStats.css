.stats {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  text-align: left;
  margin-left: auto;
  margin-right: auto;
  width: 72%;
  font-size: 14pt;
}
