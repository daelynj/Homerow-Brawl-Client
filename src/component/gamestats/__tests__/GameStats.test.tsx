import * as React from "react";
import { GameStats } from "../GameStats";
import renderer from "react-test-renderer";
import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });

const updateStats = jest.fn();
const endGame = jest.fn();

describe("GameStats", () => {
  const buildProps = (newProps = {}) => ({
    words: "this is text that is being written in a typing test".split(" "),
    finishTime: 15,
    incorrectLetters: 3,
    ID: 5,
    updateStats,
    endGame,
    statsState: {
      type: "stats",
      players: [
        {
          id: 1,
          name: "octane",
          words_typed: 11,
          wpm: 120,
          time: 7,
          mistakes: 3,
          accuracy: 86.0
        }
      ]
    },
    ...newProps
  });

  it("renders the expected game stats", () => {
    const tree = renderer.create(<GameStats {...buildProps()} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders two stat blocks", () => {
    const payload = {
      type: "stats",
      players: [
        {
          id: 1,
          name: "octane",
          words_typed: 11,
          wpm: 120,
          time: 7,
          mistakes: 3,
          accuracy: 86.0
        },
        {
          id: 2,
          name: "dominus",
          words_typed: 11,
          wpm: 60,
          time: 14,
          mistakes: 0,
          accuracy: 100.0
        }
      ]
    };

    const tree = renderer
      .create(<GameStats {...buildProps({ statsState: payload })} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
