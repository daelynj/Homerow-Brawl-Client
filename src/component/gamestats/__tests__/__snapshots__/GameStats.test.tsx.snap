// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameStats renders the expected game stats 1`] = `
<div>
  <div>
    words typed: 11
  </div>
  <div>
    WPM: 33
  </div>
  <div>
    time: 15 seconds
  </div>
  <div>
    mistakes: 3
  </div>
  <div>
    accuracy: 92.6829268292683%
  </div>
</div>
`;
